docs: Organize and expand contributing documentation

dcalhoun opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "Fixes" keyword and use "Related to" instead.
-->

Relates to p1718394299821639/1718026506.388569-slack-C06DRMD6VPZ.

## Proposed Changes

Organize and expand documentation to capture helpful guidance and information
for contributors.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

Navigate various links found within the `README` file and other documentation
files.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Have you checked for TypeScript, React or other console errors?

dcalhoun committed:
docs: Use title case consistently

The sibling headings use title case.

dcalhoun committed:
docs: Relocate contributing guidelines into separate files

Reduce the size of the `README` file and begin reorganizing code
contribution guidelines to simplify navigating additional content.

dcalhoun committed:
docs: Expand the contributing guidelines

dcalhoun committed:
docs: Expand and improve consistency of code contribution documentation

dcalhoun committed:
docs: Document CLI development tips

dcalhoun committed:
docs: Fix documentation link paths

dcalhoun committed:
docs: Fix code of conduct link path

dcalhoun committed:
docs: Fix missing punctuation

dcalhoun committed:
docs: Add links to code contribution and project setup documentation

dcalhoun committed:
fix: Improve casing consistency

dcalhoun committed:
docs: Improve documentation formatting

dcalhoun committed:
docs: Simplify copying and pasting example commands

katinthehatsite commented on docs/code-contributions.md:
````diff
@@ -0,0 +1,137 @@
+npm start
+```
+
+The app automatically launches with the Chromium developer tools opened by default. Changes to the "renderer" process code will automatically reload the app, changes to the main process code require a manual server restart or [typing `rs`](https://www.electronforge.io/cli#start) into the same terminal where the server was started.
````
Nice! TIL `rs`

katinthehatsite commented on docs/code-contributions.md:
```diff
@@ -0,0 +1,137 @@
+
+Once all required dependencies are installed[^1], you can run the app with the following command:
+
+[^1]: Whenever project dependencies are updated, you may need to run `npm install` again before running the app.
```
I am wondering if this should just be left in the body of this section instead of moving it to a footnote? I think this piece of information is quite important but could be easily missed if it is moved to the bottom of the file.

Not a strong preference though

dcalhoun replied:
Fair feedback. I relied upon footnotes as I felt this note was likely a well-understood concept for developers familiar with the JavaScript ecosystem and could be minimized as "off topic." 

However, I do not hold a strong opinion on this. I hoisted this note to higher prominence by reinstating the recommendation to always run `npm install` in https://github.com/Automattic/studio/pull/253/commits/31e357d783841f0bcd2b171b2b6222823a940863.

katinthehatsite reviewed (approved):
The changes look good. The instructions are clear and well presented and the document is well structured.

fluiddot commented on docs/code-contributions.md:
```diff
@@ -0,0 +1,137 @@
+
+### Code Formatting
+
+Code formatting is set up to make merging pull requests easier. It uses the same Prettier/ESLint mechanism as Calypso. See [JavaScript Coding Guidelines](https://github.com/Automattic/wp-calypso/blob/trunk/docs/coding-guidelines/javascript.md) for guidance on setting up your editor for code formatting.
```
I wonder if potential contributors will know what Calypso is. Maybe we could say:

```suggestion
Code formatting is set up to make merging pull requests easier. It uses the same Prettier/ESLint mechanism as Calypso, the front-end library that powers up the WordPress dashboard. See [JavaScript Coding Guidelines](https://github.com/Automattic/wp-calypso/blob/trunk/docs/coding-guidelines/javascript.md) for guidance on setting up your editor for code formatting.
```

dcalhoun replied:
Good idea! I further clarified Calypso's relationship with WordPress.com specifically in https://github.com/Automattic/studio/pull/253/commits/f015c34dda898b83b6770894d41e48002dadd607.

fluiddot reviewed (approved):
LGTM 🎊 ! Thanks @dcalhoun for improving and expanding the documentation 🙇 !

dcalhoun committed:
docs: Describe the Calypso project

Contributors may not be familiar with the Calypso project or its purpose.

Co-authored-by: Carlos Garcia <fluiddot@gmail.com>

dcalhoun committed:
docs: Further clarify Calypso's connection to WordPress.com

Calypso is not associated with the open-source WordPress project.

dcalhoun committed:
docs: Recommend running `npm install` before building the source code

Avoid potential confusion from missing or outdated dependencies by
recommending that contributors always install the latest dependencies
before building the source code.

dcalhoun commented:
Merged these changes with the feedback and approval provided thus far. 🚀 

For context and posterity, I completed this reorganization as I felt the `README` file reached a "tipping point" where it was quite verbose and unapproachable. The merged changes were inspired by [`WordPress/gutenberg`](https://github.com/WordPress/gutenberg)'s documentation. 

That said, anyone should feel welcome to further critque and/or reorganize the documentation. Documentation organization and tone is always quite subjective. I welcome others to continue proposing changes, as _any_ documentation needs to evolve alongside its project as the scope and complexity increases.
